app/models/compute_resources/foreman/model/openstack.rb

```diff
@@ -54,7 +54,28 @@ def internal_networks
       network_client.networks.all.select { |net| !net.router_external }
     end
 
+    def image_size(image_id)
+      client.get_image_details(image_id).body['image']['minDisk']
+    end
+
+    def boot_from_volume(args = {})
+      vm_name = args[:name]
+      args[:size_gb] = image_size(args[:image_ref]) if args[:size_gb].blank?
+      boot_vol = volume_client.volumes.create( { :display_name => "volume-#{vm_name}", :volumeType => "Volume", :size => args[:size_gb], :imageRef => args[:image_ref] } )
+      boot_vol_id = boot_vol.id.tr('"', '')
+      boot_vol.wait_for { status == 'available'  }
+      args[:block_device_mapping] = [ {
+        :api_ver => "v2",
+        :source_type => "volume",
+        :destination_type => "volume",
+        :delete_on_termination => "1",
+        :uuid =>boot_vol_id,
+        :boot_index => 0
+      } ]
+    end
+
     def create_vm(args = {})
+      boot_from_volume if args[:boot_from_volume] == "true"
```

Comment:
I think it's worth to add a conditional delete after the VM creation call, in case VM creation went wrong we don't want a dangling useless volume lying around :smile: